gnovm/cmd/gno/internal/fix/fix.go

```diff
@@ -0,0 +1,448 @@
+					}
+				default:
+					sc.use(n)
+				}
+
+			case *ast.TypeSpec:
+				sc.declare(n.Name, n)
+			case *ast.ValueSpec:
+				for _, name := range n.Names {
+					sc.declare(name, n)
+				}
+			case *ast.AssignStmt:
+				if n.Tok == token.DEFINE {
+					for _, name := range n.Lhs {
+						// only declare if it doesn't exist in the last scope,
+						// := allows the LHS to contain already defined values
+						// which are then simply assigned instead of declared.
+						name := name.(*ast.Ident)
+						if _, ok := sc[len(sc)-1][name.Name]; !ok {
+							sc.declare(name, n)
+						}
+					}
+				}
+			case *ast.FuncDecl:
+				id := n.Name
+				name := id.Name
+				if n.Recv != nil && len(n.Recv.List) > 0 {
+					tp := recvType(n.Recv.List[0].Type)
+					if tp != nil {
+						name = tp.Name + "." + name
+					}
+				}
+				if name != "init" {
+					sc.declare(ast.NewIdent(name), n)
+				}
+				sc.push()
+				sc.declareList(n.Recv)
+				sc.declareList(n.Type.Params)
+				sc.declareList(n.Type.Results)
+			case *ast.FuncLit:
+				sc.push()
+				sc.declareList(n.Type.Params)
+				sc.declareList(n.Type.Results)
+			case *ast.RangeStmt:
+				sc.push()
+				if n.Tok == token.DEFINE {
+					if id, ok := n.Key.(*ast.Ident); ok {
+						sc.declare(id, n)
+					}
+					if id, ok := n.Value.(*ast.Ident); ok {
+						sc.declare(id, n)
+					}
+				}
+			case *ast.BlockStmt,
+				*ast.IfStmt,
+				*ast.SwitchStmt,
+				*ast.TypeSwitchStmt,
+				*ast.CaseClause,
+				*ast.CommClause,
+				*ast.ForStmt,
+				*ast.SelectStmt,
+				*ast.File:
+				sc.push()
+			}
+			return true
+		},
+		func(c *astutil.Cursor) bool {
+			if post != nil && !post(c, sc) {
+				return false
+			}
+
+			if isBlockNode(c.Node()) {
+				sc.pop()
+			}
+
+			return true
+		},
+	)
+}
+
+func isBlockNode(n ast.Node) bool {
+	switch n.(type) {
+	case *ast.FuncDecl,
+		*ast.BlockStmt,
+		*ast.FuncLit,
+		*ast.IfStmt,
+		*ast.SwitchStmt,
+		*ast.TypeSwitchStmt,
+		*ast.CaseClause,
+		*ast.CommClause,
+		*ast.ForStmt,
+		*ast.SelectStmt,
+		*ast.RangeStmt,
+		*ast.File:
+		return true
+	}
+	return false
+}
+
+type definitionUsages struct {
+	def    ast.Node
+	usages []*ast.Ident
+}
+
+func (du definitionUsages) rename(name string) {
+	for _, us := range du.usages {
+		us.Name = name
+	}
+}
+
+type scope map[string]*definitionUsages
+
+type scopes []scope
+
+func (s scopes) lookup(name string) ast.Node {
+	for _, scope := range slices.Backward(s) {
+		if stmt, ok := scope[name]; ok {
+			return stmt.def
+		}
+	}
+	return nil
+}
+
+func (s *scopes) push() {
+	*s = append(*s, scope{})
+}
+
+func (s *scopes) pop() {
+	*s = (*s)[:len(*s)-1]
+}
+
+func (s scopes) declare(name *ast.Ident, stmt ast.Node) {
+	if name.Name == "_" {
+		return
+	}
+	sc := s[len(s)-1]
+	if du, ok := sc[name.Name]; ok {
+		if du.def != nil {
+			panic(fmt.Sprintf("duplicate declaration of ident %q", name))
+		}
+		// This name was encountered before, but the name had not been declared yet.
+		du.def = stmt
```

Comment:
is this being overwritten on the next line ?

Reply:
good catch! added a return and a regression test for the bug it creates :)